kernel/src/syscall.rs

```diff
@@ -522,111 +552,197 @@ impl SyscallReturn {
     /// Encode the system call return value into 4 registers, following the
     /// encoding specified in TRD104. Architectures which do not follow TRD104
     /// are free to define their own encoding.
+    /// TODO: deprecate in favour of the more general one
     pub fn encode_syscall_return(&self, a0: &mut u32, a1: &mut u32, a2: &mut u32, a3: &mut u32) {
+        if core::mem::size_of::<MetaPtr>() == core::mem::size_of::<u32>() {
+            // SAFETY: if the two unsized integers are the same size references to them
+            // can be safely transmuted.
+            // Ugly coercion could be avoided by first copying to the stack, then assigning with
+            // "as" in order to satisfy the compiler. But I expect this function will disappear
+            // in favour of just using the usize one.
+            unsafe {
+                let a0 = &mut *(core::ptr::from_mut(a0) as *mut MetaPtr);
+                let a1 = &mut *(core::ptr::from_mut(a1) as *mut MetaPtr);
+                let a2 = &mut *(core::ptr::from_mut(a2) as *mut MetaPtr);
+                let a3 = &mut *(core::ptr::from_mut(a3) as *mut MetaPtr);
+                self.encode_syscall_return_mptr(a0, a1, a2, a3);
+            }
+        } else {
+            panic!("encode_syscall_return used on a 64-bit platform or CHERI platform")
```

Comment:
I think this needs to remain a compile time check. This is very low level to have a panic.

Reply:
My logic was that people would therefore find out very fast that something was wrong and they were calling the wrong function.

I will check that `const _ : () = ....` does what we want here. I am just worried that the existence of the function on 32-bit platforms in, say, debug builds would cause the compiler error.

Reply:
Checked. The static assert will trigger in an undesirable way on 64-bit platforms. Its not really an error for this function to exist, just use it.

Unless we use cfg[] to remove this function entirely, I think it is correct to just panic on 64-bit systems.

Reply:
Any reason this shouldn't be an `assert!` instead?

For @bradjc I _believe_ (and we can confirm) that because this can be resolved at compile time (both sides of the `==` are const), the panic/assert are going to be elided at compile time on 32-bit platforms.

@LawrenceEsswood generally we avoid resolving comments directed to us unless it becomes irrelevant (which GitHub does automatically when the code changes enough), and instead the author of the comment (or potentially some other moderator) should resolve when their question is answered

Reply:
Sure this, can be assert too. Will roll into the next round of patches (and leave this open!).

Reply:
Echoing the comment from https://github.com/tock/tock/pull/4174/files/7ec6bdd51152cc054c2f68dcab3425928ad11876#r1835188434. This ought to not be platform specific; it should be callable on 64-bit platforms as well (even if just for testing). It's fine for this to be less efficient on those platforms (such as allocating another set of pointers on the stack and copying, instead of these casts).